pkg/collector/corechecks/cluster/ksm/kubernetes_state.go

```diff
@@ -771,21 +777,35 @@ func (k *KSMCheck) getClusterName() {
 	}
 }
 
-// initTags avoids keeping a nil Tags field in the check instance
-// Sets the kube_cluster_name tag for all metrics.
-// Adds the global user-defined tags from the Agent config.
-func (k *KSMCheck) initTags() {
-	if k.instance.Tags == nil {
-		k.instance.Tags = []string{}
-	}
-
+// initTags adds the following tags to every metric:
+// - `kube_cluster_name` tag for all metrics.
+// - the global user-defined tags from the Agent config.
+func (k *KSMCheck) initTags() error {
+	// Add the cluster name tag to the global tags
+	globalTags := []string{}
 	if k.clusterNameTagValue != "" {
-		k.instance.Tags = append(k.instance.Tags, "kube_cluster_name:"+k.clusterNameTagValue)
+		globalTags = append(globalTags, "kube_cluster_name:"+k.clusterNameTagValue)
 	}
-
+	// Add configured tags to the global tags
 	if !k.instance.DisableGlobalTags {
-		k.instance.Tags = append(k.instance.Tags, configUtils.GetConfiguredTags(k.agentConfig, false)...)
+		globalTags = append(globalTags, configUtils.GetConfiguredTags(k.agentConfig, false)...)
+	}
+
+	// Get the current check tags
+	sender, err := k.GetRawSender()
+	if err != nil {
+		return err
 	}
+	checkTags := sender.GetCheckCustomTags()
+
+	// Merge the global tags with the check tags
+	newCheckTags := append(checkTags, globalTags...)
+	// Pre-sort them to avoid the overhead of InsertionSort in the aggregator
+	newCheckTags = sort.UniqInPlace(newCheckTags)
```

Comment:
IMO,
here you need to do a copy 

```suggestion
    
	checkTags := sender.GetCheckCustomTags()
	newCheckTags := make([]string, len(checkTags))
	copy(newCheckTags, checkTags)

	// Merge the global tags with the check tags
	newCheckTags := append(newCheckTags, globalTags...)
	// Pre-sort them to avoid the overhead of InsertionSort in the aggregator
	newCheckTags = sort.UniqInPlace(newCheckTags)
```

Reply:
I think that even if we don't copy the tags it should work as we basically do `Get(key)` and `Set(key, value)` with the same key.